sui_core/src/authority/authority_store.rs

```diff
@@ -710,40 +622,52 @@ impl<const ALL_OBJ_VER: bool, S: Eq + Serialize + for<'de> Deserialize<'de>>
                 .map(|(object_id, (_, new_object))| (object_id, new_object)),
         )?;
 
-        // Update the indexes of the objects written
-
-        // This is the critical region: testing the locks and writing the
-        // new locks must be atomic, and no writes should happen in between.
-        {
-            // Acquire the lock to ensure no one else writes when we are in here.
-            let _mutexes = self.acquire_locks(&active_inputs[..]);
-
-            // Check the locks are still active
-            // TODO: maybe we could just check if the certificate is there instead?
-            let locks = self.transaction_lock.multi_get(&active_inputs[..])?;
-            for object_lock in locks {
-                object_lock.ok_or(SuiError::TransactionLockDoesNotExist)?;
-            }
+        // There used to be a critical region here.  This is no longer necessary.
+        // 1) Atomic lock changes are now handled by a lockservice
+        // 2) The presence of a certificate guarantees that locks have been held by >2f+1 authorities.
+        // 3) Lock updates are written after objects are inserted. This guarantees objects exist
+        //    before we allow others to transact on the objects.
+        // If we really want to be careful we can issue a request to check the locks again.
+        // Note that the old lock checking logic was faulty, it doesn't check the locks belong to
```

Comment:
The check of the locks as part of the critical region here was not needed to protect the safety of locks but to prevent a time-of-check time-of-use race conditions between (1) reading the objects and checking their version is correct to apply this certificate, and then (2) actually applying the certificate.

We need to ensure here that the object versions we have in the DB are all in the version that this transaction requires. Otherwise, it might be the case that we re-create them by mistake. Here is a trace that would lead to this issue:

Req1 is received applying Cert1
Req2 is received applying Cert1 (same as above)
Req3 is received applying Cert2
Req1 check the objects and they in a good version, so proceeds to execute Cert1
(#1) Re2 checks the (same) objects and they are also in a good version so proceeds with (same) Cert1
Req1 terminates and updates the objects on disk.
Req3 now checks the objects, which we generated by Cert1/Req1 they exist, and proeeds and updates them
(#2) Req2 now resumes computes (again) Cert1, and re-writes the objects at an earlier version (<- BUG now we have reverted objects to an older version)

The issue with the above trace is that check happens in #1 and use and update happens in #2 and by the time #2 happens #1 may not be true any more.

Morality of the story: we need here a critical region that checks again the version of all objects, ensures they are correct, and atomically updates them. Unless we keep all versions of objects at which point we can do away with this, since a write does not overwrite the "latest" object.

Reply:
The alternative is to do all this in a lock free way through the definition of a merge operator on object that simply keeps the highest version written. This is also cool.

Reply:
@gdanezis I would agree that a lock free design (I would argue this is not a lock free design, but a critical region free design) does require a certain degree of idempotency, that is, that you can allow certain operations to be repeated and repeated writes must be safe, so long as the lock itself is atomic.     And the object storage currently is not idempotent, because it is keyed by ObjID.   There are some problems with the ObjID key actually, so this likely should be changed.    However don't have the space to go into it here. 
 
The core problem is that normally acquiring the lock is supposed to give you some guarantee, that it is safe to operate on those objects.  You should not have to check the lock again or use the critical region.  The problem is that due to the key of the objects table, this guarantee does not exist today.

Checking the locks before writing the batch gives you almost as good of that guarantee.  Even without the critical region, the only way this would fail is if Req2 proceeds, computes the computation, checks the locks and they exist, but at that moment somehow is suspended, req3 proceeds and does the full computation, then req2 resumes and overwrites.  But possible.

For now I'll restore the critical region and fully document the problems here, but there are at least two solutions:

1. Change the object model to inherently support multiple versions.  This is actually needed for other reasons - proper caching if/when we move to distributed object stores, smaller and more efficient object references in transactions, and just much less mutability and safety (#218)
2. Change the object writing to be a "write object at version Y IFF it is currently at version X" - a compare and swap operation.  This could be implemented with a merge operator, and is safer than simply one which just gives you the higher version.

I'm going to comment on the other issue though.

Reply:
Yes, I agree that #218 goes some way to solve this issue BUT only for objects themselves. The various indexes we maintain might still need to be updated in a causal order. You can check that when we come to it.